# Introduction to SQL Using Python: Aggregating Data with the GROUP BY Statement

In my last blog, I discussed how to filer data using SQL with the __WHERE__ statement. This blog is a tutorial on how to aggregate data using the __GROUP BY__ statement.

In [1]:
# Import necessary libraries

import sqlite3
import pandas as pd

In [2]:
# Connect to database
conn = sqlite3.connect('''database.sqlite''')

# Create cursor object
cur = conn.cursor()

In [ ]:
cur.execute('''Enter SQL query here;''') # Runs SQL query
data = pd.DataFrame(cur.fetchall()) # Converts SQL query results into dataframe format
data.columns = [x[0] for x in cur.description] # Labels the columns of the dataframe
data # View SQL results dataframe

In [7]:
# View Teams dataframe

cur.execute('''SELECT * 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season        TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0     2017   Bayern Munich        27         26                 15   
1     2017        Dortmund        33         25                 18   
2     2017      Leverkusen        31         24                 15   
3     2017      RB Leipzig        30         23                 15   
4     2017      Schalke 04        29         24                 17   
5     2017      M'gladbach        31         25                 17   
6     2017       Wolfsburg        31         24                 14   
7     2017         FC Koln        24         26                  9   
8     2017      Hoffenheim        31         24                 14   
9     2017          Hertha        26         26                 12   
10    2017           Mainz        31         25                 16   
11    2017         Hamburg        31         25                 15   
12    2017   Werder Bremen        32         26                 19   
13    2017   Ein Frankfurt        34         24                 18   
14    2017        Augsburg        36         26                 20   
15    2017        Freiburg        26         26                  8   
16    2017       Stuttgart        24         24                 16   
17    2017        Hannover        31         25                 13   
18    2016   Bayern Munich        33         25                 18   
19    2016        Dortmund        32         24                 17   
20    2016      Leverkusen        39         23                 21   
21    2016      Schalke 04        39         24                 22   
22    2016       Wolfsburg        43         24                 21   
23    2016      M'gladbach        33         23                 15   
24    2016         Hamburg        38         24                 19   
25    2016           Mainz        39         23                 22   
26    2016         FC Koln        30         24                 13   
27    2016      Hoffenheim        33         24                 15   
28    2016          Hertha        33         24                 14   
29    2016   Werder Bremen        41         25                 19   
..     ...             ...       ...        ...                ...   
438   2006  Greuther Furth        29         24                 14   
439   2006         CZ Jena        37         25                 11   
440   2006     Munich 1860        26         24                 11   
441   2006        RW Essen        27         26                 11   
442   2006      Burghausen        28         25                 13   
443   2006  Erzgebirge Aue        26         27                 11   
444   2006       Offenbach        30         27                  9   
445   2006       Paderborn        28         25                 12   
446   2006    Unterhaching        31         26                 12   
447   2006    Braunschweig        38         26                 14   
448   2006        Augsburg        25         27                  7   
449   2006         Koblenz        28         26                 11   
450   2005     Munich 1860        31         25                 15   
451   2005          Bochum        31         25                 16   
452   2005   Hansa Rostock        31         25                 12   
453   2005        Freiburg        25         24                 14   
454   2005          Aachen        25         27                  6   
455   2005         Dresden        33         26                 13   
456   2005       Karlsruhe        27         25                  7   
457   2005  Greuther Furth        30         25                 14   
458   2005     Saarbrucken        31         26                 15   
459   2005           Ahlen        24         28                  9   
460   2005         Cottbus        26         25                 12   
461   2005  Erzgebirge Aue        25         27                  8   
462   2005      Burghausen        28         25                  9   
463   2005    

# COUNT( )

The first aggregation function we will look at is the __COUNT( )__ function. Previously we have either put a single asterisk in the __SELECT__ statment or we had listed the names of the columns we wanted to be returned. The __COUNT( )__ function can also be listed in the __SELECT__ statement and will count the number of rows specified inbetween the parentheses. The query below has the function __COUNT(*)__ in the __SELECT__ statement. __COUNT(*)__ will count all the rows from the table specified. The below query counts all the rows in the __Teams__ dataset. 

In [9]:
# View the number of rows from Teams dataframe

cur.execute('''SELECT COUNT(*) 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

COUNT(*)
0       468

The are a total of 468 rows included in the __Teams__ dataset. What happens if we insert a column name instead of an asterisk between the parentheses of the __COUNT( )__ function, for example, __COUNT(Season)__? The function will then count all the rows in the __Season__ column where there is a non-null value. If a value of __Season__ is null for any row, then that row will not be counted. If there are no null values in the __Season__ column then the following function will return 468, the same number as the total number of rows in the __Teams__ dataset.

In [10]:
# Count the number of Season rows in the Teams Dataframe

cur.execute('''SELECT COUNT(Season) 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

COUNT(Season)
0            468

Great, the number of rows with non-null values is 468, the same as the total amount of rows included in the __Teams__ dataset. That means there are no null values. To practice using the __COUNT( )__ statement, count the number of rows in the __Matches__ dataset and compare your query to the one below:

In [11]:
# Count the number of rows in the Matches dataframe

cur.execute('''SELECT COUNT(*) 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT(*)
0     24625

There should be 24,625 rows in the __Matches__ dataset. Check to see if there are any missing values in the column __Season__ from the __Matches__ dataset and compare your query to the one below:

In [12]:
# Count the number of Season rows in the Matches Dataframe

cur.execute('''SELECT COUNT(Season) 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT(Season)
0          24625

There are 24,625 rows with non-null values in the __Season__ column in the __Matches__ dataset. Since this is the same as the total number of rows in the __Matches__ dataset, there are no missing/null values in the __Season__ column.

# SELECT DISTINCT

With the __COUNT( )__ function, we were able to see how many rows were listed in both the __Teams__ and __Matches__ datasets. We also saw the number of rows that did not have null values in the __Seasons__ column for both datasets. What if we want to know the unqiue values in the __Seasons__ column in the __Teams__ dataset. We can't use __COUNT(Season)__, because that just gives us the number of rows that have a specified value in the __Season__ column. Instead we will use the __DISTINCT__ clause in the __SELECT__ statement. The below query will show us all of unqiue values that are included in the __Season__ column in the __Teams__ dataset.

In [13]:
# View the unqiue Season values in the Teams Dataframe

cur.execute('''SELECT DISTINCT Season
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season
0    2017
1    2016
2    2015
3    2014
4    2013
5    2012
6    2011
7    2010
8    2009
9    2008
10   2007
11   2006
12   2005

You can also use more than one column name when using the __DISTINCT__ clause is the __SELECT__ statement. For example, one soccer team may be the home team multiple times in a season, but if we only want to know the seasons each team was a home team we can use the following query:

In [19]:
cur.execute('''SELECT DISTINCT Season, HomeTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Season            HomeTeam
0       2009          Oberhausen
1       2009         Munich 1860
2       2009       Frankfurt FSV
3       2009               Ahlen
4       2009        Union Berlin
5       2009           Paderborn
6       2009           Bielefeld
7       2009      Kaiserslautern
8       2009       Hansa Rostock
9       2009      Greuther Furth
10      2009             Koblenz
11      2009             Cottbus
12      2009            Duisburg
13      2009  Fortuna Dusseldorf
14      2009              Aachen
15      2009            Augsburg
16      2009            St Pauli
17      2009              Bochum
18      2009       Bayern Munich
19      2009          M'gladbach
20      2009               Mainz
21      2009             Hamburg
22      2009           Stuttgart
23      2009            Dortmund
24      2009          Schalke 04
25      2009            Freiburg
26      2009       Werder Bremen
27      2009              Hertha
28      2009            Hannover
29      2009       Ein Frankfurt
...      ...                 ...
1356    2015         Aston Villa
1357    2015         Southampton
1358    2015          Sunderland
1359    2015             Swansea
1360    2015           Tottenham
1361    2015             Watford
1362    2015            West Ham
1363    2015      Crystal Palace
1364    2015            Man City
1365    2015           Liverpool
1366    2016             Burnley
1367    2016      Crystal Palace
1368    2016             Everton
1369    2016                Hull
1370    2016            Man City
1371    2016       Middlesbrough
1372    2016         Southampton
1373    2016             Arsenal
1374    2016         Bournemouth
1375    2016             Chelsea
1376    2016          Man United
1377    2016           Leicester
1378    2016               Stoke
1379    2016             Swansea
1380    2016           Tottenham
1381    2016             Watford
1382    2016           West Brom
1383    2016          Sunderland
1384    2016            West Ham
1385    2016           Liverpool

[1386 rows x 2 columns]

The above results show every distinct __HomeTeam__ that was in every distinct __Season__ from the __Matches__ table. This might seem a bit confusing right now but we will play around more with the __DISTINCT__ clause later. For now, practice selecting all the unique __AwayTeams__ there are in the __Matches__ dataset and compare your query to the one below:

In [22]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT DISTINCT AwayTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

AwayTeam
0        Kaiserslautern
1             Karlsruhe
2            Leverkusen
3              Nurnberg
4            Schalke 04
5             Stuttgart
6         Werder Bremen
7                Bochum
8              Hannover
9         Hansa Rostock
10               Hertha
11                Mainz
12          Munich 1860
13             St Pauli
14            Wolfsburg
15           M'gladbach
16            Bielefeld
17              Cottbus
18       Greuther Furth
19           Oberhausen
20               Aachen
21                Ahlen
22         Union Berlin
23            Paderborn
24             Augsburg
25              Koblenz
26           Hoffenheim
27        Frankfurt FSV
28   Fortuna Dusseldorf
29        Bayern Munich
..                  ...
98     Sheffield United
99            Liverpool
100         Aston Villa
101              Oldham
102       Nott'm Forest
103           Leicester
104      Crystal Palace
105              Bolton
106       Middlesbrough
107          Sunderland
108               Derby
109            Barnsley
110            Charlton
111            Bradford
112             Watford
113              Fulham
114           West Brom
115         Middlesboro
116          Birmingham
117              Wolves
118          Portsmouth
119               Wigan
120             Reading
121               Stoke
122                Hull
123             Burnley
124           Blackpool
125             Swansea
126             Cardiff
127         Bournemouth

[128 rows x 1 columns]

# Using COUNT( ) & DISTINCT Together

We just saw how to get all the unique __AwayTeam__ names from the Matches dataset above. What happens when we want to know how many unqiue values are returned? to get the count of the unique __AwayTeam__ names, we can use the __COUNT( )__ function together with the __DISTINCT__ clause. The below query will find all the the distinct AwayTeam names in the __Matches__ dataset and return the count or number of rows that are included in distinct AwayTeam names.

In [23]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT COUNT (DISTINCT AwayTeam)
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

COUNT (DISTINCT AwayTeam)
0                        128

Now we can see all that there are 128 unique Away Teams in the __Matches__ dataset. We can also see that the name of this column is named __COUNT (DISTINCT AwayTeam)__, this is a bit long and unclear. We can change the name of a column using an __AS__ clause. In the __SELECT__ statement, we can rename any column just by writing __AS__ after the data we are selecting and giving it a new name. The below query shows the number of unique away teams in the __Matches__ dataset but this time, the column header is __Num_of_AwayTeams__.

In [25]:
# Show all the unqiue AwayTeams in the Matches Dataframe

cur.execute('''SELECT COUNT (DISTINCT AwayTeam) AS Num_of_AwayTeams
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Num_of_AwayTeams
0               128

To practice using the __COUNT( )__ function and __DISTINCT__ clause together, query the number of distinct __Seasons__ included in the __Teams__ dataset and rename the column to __Num_of_Seasons__. Compare your query to the one below:

In [27]:
# Count of unqiue seasons in the Teams dataset

cur.execute('''SELECT COUNT(DISTINCT Season) AS Num_of_Seasons
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Num_of_Seasons
0              13

# MAX( ) & MIN( ) Functions

We will continue to look at how data can be aggregated in the __SELECT__ statement before we get to the __GROUP BY__ statement. Next we will look at how to use the __MAX( )__ function and __MIN( )__ function. Perhaps we want to know what is the maximum stadium capacity is that is included in the __Teams__ dataset. We can use the query below:

In [29]:
# Return the maximum StadiumCapacity in the Teams datas

cur.execute('''SELECT MAX(StadiumCapacity)
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

MAX(StadiumCapacity)
0                81359

In [28]:
# View Teams dataframe

cur.execute('''SELECT * 
               FROM Teams;''')
Teams_df =pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df

Season        TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0     2017   Bayern Munich        27         26                 15   
1     2017        Dortmund        33         25                 18   
2     2017      Leverkusen        31         24                 15   
3     2017      RB Leipzig        30         23                 15   
4     2017      Schalke 04        29         24                 17   
5     2017      M'gladbach        31         25                 17   
6     2017       Wolfsburg        31         24                 14   
7     2017         FC Koln        24         26                  9   
8     2017      Hoffenheim        31         24                 14   
9     2017          Hertha        26         26                 12   
10    2017           Mainz        31         25                 16   
11    2017         Hamburg        31         25                 15   
12    2017   Werder Bremen        32         26                 19   
13    2017   Ein Frankfurt        34         24                 18   
14    2017        Augsburg        36         26                 20   
15    2017        Freiburg        26         26                  8   
16    2017       Stuttgart        24         24                 16   
17    2017        Hannover        31         25                 13   
18    2016   Bayern Munich        33         25                 18   
19    2016        Dortmund        32         24                 17   
20    2016      Leverkusen        39         23                 21   
21    2016      Schalke 04        39         24                 22   
22    2016       Wolfsburg        43         24                 21   
23    2016      M'gladbach        33         23                 15   
24    2016         Hamburg        38         24                 19   
25    2016           Mainz        39         23                 22   
26    2016         FC Koln        30         24                 13   
27    2016      Hoffenheim        33         24                 15   
28    2016          Hertha        33         24                 14   
29    2016   Werder Bremen        41         25                 19   
..     ...             ...       ...        ...                ...   
438   2006  Greuther Furth        29         24                 14   
439   2006         CZ Jena        37         25                 11   
440   2006     Munich 1860        26         24                 11   
441   2006        RW Essen        27         26                 11   
442   2006      Burghausen        28         25                 13   
443   2006  Erzgebirge Aue        26         27                 11   
444   2006       Offenbach        30         27                  9   
445   2006       Paderborn        28         25                 12   
446   2006    Unterhaching        31         26                 12   
447   2006    Braunschweig        38         26                 14   
448   2006        Augsburg        25         27                  7   
449   2006         Koblenz        28         26                 11   
450   2005     Munich 1860        31         25                 15   
451   2005          Bochum        31         25                 16   
452   2005   Hansa Rostock        31         25                 12   
453   2005        Freiburg        25         24                 14   
454   2005          Aachen        25         27                  6   
455   2005         Dresden        33         26                 13   
456   2005       Karlsruhe        27         25                  7   
457   2005  Greuther Furth        30         25                 14   
458   2005     Saarbrucken        31         26                 15   
459   2005           Ahlen        24         28                  9   
460   2005         Cottbus        26         25                 12   
461   2005  Erzgebirge Aue        25         27                  8   
462   2005      Burghausen        28         25                  9   
463   2005    

# ORDER BY Statement

In [ ]:
# DESC & ASC